File: javascript/tests/credential.test.js
/**
 * Tests for credential issuance and validation.
 */

import { test, describe } from 'node:test';
import assert from 'node:assert';
import { generateKeyPair } from '../src/crypto.js';
import { verifyJwt } from '../src/jwt.js';
import { Capability } from '../src/capability.js';
import { issueCredential, validateCredentialAgainstDiscovery } from '../src/credential.js';

describe('issueCredential', () => {
    test('issues and verifies credential', () => {
        const kp = generateKeyPair();

        const jwt = issueCredential(
            kp.privateKeyPem,
            'test-2026-01',
            'example.com',
            'urn:agentpin:example.com:agent',
            'verifier.com',
            [new Capability('read:data')],
            null,
            null,
            3600
        );

        const { header, payload } = verifyJwt(jwt, kp.publicKeyPem);
        assert.strictEqual(header.kid, 'test-2026-01');
        assert.strictEqual(payload.iss, 'example.com');
        assert.strictEqual(payload.sub, 'urn:agentpin:example.com:agent');
        assert.strictEqual(payload.aud, 'verifier.com');
        assert.strictEqual(payload.agentpin_version, '0.1');
        assert.ok(payload.exp > payload.iat);
    });
});

describe('validateCredentialAgainstDiscovery', () => {
    test('valid subset passes', () => {
        const discCaps = ['read:*', 'write:report'];
        const credCaps = ['read:data'];
        assert.doesNotThrow(() => validateCredentialAgainstDiscovery(credCaps, discCaps));
    });

    test('exceeding capabilities throws', () => {
        const discCaps = ['read:data'];
        const credCaps = ['delete:data'];
        assert.throws(
            () => validateCredentialAgainstDiscovery(credCaps, discCaps),
            /exceed/i
        );
    });
});

File: go/pkg/credential/credential_test.go
package credential

import (
	"testing"

	"github.com/ThirdKeyAi/agentpin/go/pkg/crypto"
	"github.com/ThirdKeyAi/agentpin/go/pkg/jwt"
	"github.com/ThirdKeyAi/agentpin/go/pkg/types"
)

func TestIssueCredential(t *testing.T) {
	kp, _ := crypto.GenerateKeyPair()
	priv, _ := crypto.LoadPrivateKey(kp.PrivateKeyPEM)
	pub, _ := crypto.LoadPublicKey(kp.PublicKeyPEM)

	cred, err := IssueCredential(
		priv, "test-2026-01", "example.com", "urn:agentpin:example.com:agent",
		"verifier.com",
		[]types.Capability{"read:data"},
		nil, nil, 3600,
	)
	if err != nil {
		t.Fatal(err)
	}
	header, payload, err := jwt.VerifyJWT(cred, pub)
	if err != nil {
		t.Fatal(err)
	}
	if header.Kid != "test-2026-01" {
		t.Fatalf("kid: %s", header.Kid)
	}
	if payload.Iss != "example.com" || payload.Sub != "urn:agentpin:example.com:agent" {
		t.Fatal("iss/sub mismatch")
	}
	if payload.Aud != "verifier.com" || payload.AgentpinVersion != "0.1" {
		t.Fatal("aud/version mismatch")
	}
	if payload.Exp <= payload.Iat {
		t.Fatal("exp must be > iat")
	}
}

func TestValidateCredentialAgainstDiscovery(t *testing.T) {
	disc := []types.Capability{"read:*", "write:report"}
	if err := ValidateCredentialAgainstDiscovery([]types.Capability{"read:data"}, disc); err != nil {
		t.Fatalf("subset should pass: %v", err)
	}
	if err := ValidateCredentialAgainstDiscovery([]types.Capability{"delete:data"}, []types.Capability{"read:data"}); err == nil {
		t.Fatal("delete > read should fail")
	}
}
